### DISCUSSION TITLE
Shaders in Stride

### DISCUSSION THREAD

**Discussion by zver322** [stride3d/stride#1716] [Q&A]
Hello, everyone! Mt name is Lev and I'm started leraning how to work with SDSL and shaders in Stride in general. 
I have several question because it's not clearly said in documentaion about some aspects of that language and how to write different typs of shaders.
There are my questions:
1) I found that this shader language add new keywords
![image](https://github.com/stride3d/stride/assets/84985250/fe3095dc-f42c-4a8d-83d5-4e5adecdc277)
And I don't know what attributes and methods are inside them. For example, I found in one git repository code for motion blur that I can use streams.TexCoord. If anoubody know where I can find a good explanation of what inside this classes? In the main page it's not clear about what I can do with them.
Link for git repo with motion blur: https://github.com/ykafia/UnrealMotion/blob/master/UnrealMotion/Effects/VelocityTile.sdsl
The main Stride page: https://doc.stride3d.net/4.0/en/manual/graphics/effects-and-shaders/shading-language/index.html
2)  If you have some blur shader which were written on this shader language it will could be great. Especially I'm trying to find Gaussian Blur. There are no website which shows how to use this language and I only can find info in git repositories where it is not clear how to use them in Stride.
3) Does someone know how to write Compute Shaders because there only small amout of info on the main page. 
![image](https://github.com/stride3d/stride/assets/84985250/ac649f0c-67cf-42bf-b436-6c17e67a5213)

**Comment by tebjan** [comment]
Did you find the [Stride Shader explorer](https://github.com/tebjan/Stride.ShaderExplorer)? It gives you an overview of the shaders and how they inherit from each other. This is a great learning source. Make sure Stride is installed before starting it.

Also, have a look at the shader examples in the Stride launcher. Create the project and inspect it.

Here are many many image shaders:
https://github.com/bj-rn/VL.Addons/tree/main/vl/shaders

And here are many more shaders in general:
https://github.com/vvvv/VL.StandardLibs/tree/main/VL.Stride.Runtime/src/Effects

But Stride also comes with Gaussian blur, I think. The main issue for you will be that most of your examples are multi-pass shaders and you will need to write CPU code to manage the shader passes. That will require an understanding of the Stride render pipeline.

If you just want to learn how to write shaders in SDSL you can [try vvvv](https://visualprogramming.net/) where everything is interactive and in real-time. If you then have a shader you like, you can integrate it into your Stride game. Have a look at the help patch "How to write a shader" in the vvvv help browser.

**Reply by zver322** [reply]
Thank you, your answer gave me a direction where to move and explore. If I have any quesations about shaders can I ask you here?

**Reply by zver322** [reply]
Hello again, sorry for disturbing you.
I had a problem with Stride and shader which I copied and modifed for Stride Shader Exporer.
I wrote a shader of Gaussian Blur which firstly looked like this:
![image](https://github.com/stride3d/stride/assets/84985250/748317f5-abdd-41da-867c-5098bb86662f)
Then I modified it, because I don't clearly understood some moments and sdsl.cs file incrorrecly tried to create array of float2
![image](https://github.com/stride3d/stride/assets/84985250/1de90633-25f1-4b36-aa62-ed383a7b5c36)
I was inspired by this motion blur code and tried to write the same structure of code for affecting on image
Link: https://github.com/ykafia/UnrealMotion/blob/master/UnrealMotion/PostEffect/MotionBlur.cs
In my project I got a texture and tried to use ImageEffectShader class. My code looks like that.
![image](https://github.com/stride3d/stride/assets/84985250/a76e1935-914a-43da-830d-c2b7ccc1643a)
In Stride Game Studio I see such problem 
![image](https://github.com/stride3d/stride/assets/84985250/7344e146-25ad-4dd9-823f-25f7e38f8e6d)
![image](https://github.com/stride3d/stride/assets/84985250/cacea82a-ab75-4c79-acca-0b46e1baee93)
I found that proble is with the first line and I don't know how to fix it. I created another shader which just returned float4(1, 0, 0, 0) and it throwed me the same error. Maybe I made stupid mistake or I don't understand some hidden stones of how to use ImageEffectShader. 
And if you know the answer can you also explain what means this line of error: "Error: One or more errors occurred. (Could not compile shader. See error messages.[(0,0)]: Error:  E1202: The mixin [ShaderEffects/HorizontalBlur] in [Mix0] dependency is not in the module"

**Reply by tebjan** [reply]
All shaders must be in the folder `/Effects` or if you want to use custom folders you have to add them to the *.sdpkg file, otherwise the asset compiler cannot find the file.

**Reply by zver322** [reply]
Thank you.
But I got the same error afer I created a folder. I created this folder in code folder.
![image](https://github.com/stride3d/stride/assets/84985250/c3d58d71-5850-429a-ae4d-21acd63bad86)
![image](https://github.com/stride3d/stride/assets/84985250/b6f3be4b-0e7f-43be-a143-8d7c344d503d)
![image](https://github.com/stride3d/stride/assets/84985250/01e9198e-e025-464c-976a-a485f2121842)

**Reply by Doprez** [reply]
I personally dont fully understand shaders but I have used this one before in my projects [outline shader](https://github.com/herocrab/StrideMeshOutlineRenderFeature/blob/main/MeshOutlineRenderFeature/MeshOutlineRenderFeature/Assets/Effects/MeshOutlineShader.sdsl) hopefully there is something that can be compared to what you need.

[Awesome-stride](https://github.com/Doprez/Awesome-Stride#example-projects) this is a repo I made to give a bunch of examples, maybe there is also something in the projects section that can help.
